File: src/planner.ts
import { execFileSync } from "node:child_process";

/**
 * Planning tier — the insight/execution split.
 *
 * The thesis: a task needs a capable model to *decide what to do*, but the doing
 * is mechanical. So spend frontier tokens once on a plan, then execute the steps
 * on a cheap local model. Routing (pick one model for the whole task) cannot
 * capture that; decomposition can.
 *
 * This is also the shape small models actually succeed at. Measured earlier: a
 * 1.2B refused a politely-worded task and complied with an imperative one, and a
 * 12B produced empty patches on tasks it could not scope. "In file X, change Y
 * to Z" is a different problem from "fix this bug".
 *
 * The planning model may be anything the user has access to — a frontier model
 * they pay for, or another local one. It is a separate endpoint from execution,
 * so `--provider anthropic` for planning and LM Studio for execution is normal.
 */
export interface PlanStep {
  /** File the step touches, when the planner can name one. */
  file?: string;
  /** One imperative instruction, scoped small enough for a cheap model. */
  action: string;
}

export interface Plan {
  steps: PlanStep[];
  /** Anything the executor needs to know that is not a step. */
  notes?: string;
  tokensIn: number;
  tokensOut: number;
  model: string;
}

const PLAN_SYSTEM = `You are planning work for a SMALL, LITERAL coding model that will execute your steps.
It has file read/write/edit tools and a shell. It is not clever: it follows instructions exactly and
fails when a step requires judgement or spans several files at once.

Produce a plan as JSON only, no prose:
{"steps":[{"file":"path/to/file","action":"one imperative instruction"}],"notes":"optional context"}

Rules for steps:
- Each step must be independently executable and verifiable.
- Name the exact file. Name the exact change. Quote the literal text to find and the literal text to
  replace it with wherever you can.
- No step may require deciding *whether* to do something. You decide; the executor acts.
- Prefer 1-6 steps. If the task genuinely needs more, the task is too large — say so in notes.
- If a step needs a command run to verify, make that its own step with the exact command.

NEVER emit an investigation step. "Read X to understand it", "examine the structure", "review the
code", "analyse the current implementation" and "locate the bug" are all forbidden: the executor
reads files on its own initiative, so a step like that costs a planning round trip and decides
nothing. Every step must change something or verify something. If you cannot say what to change
without reading the file, you were given enough context to try anyway — make your best concrete
call and note the assumption in "notes".

A plan whose steps are all investigation is worse than no plan, because the executor pays for it
twice: once to generate and once to carry in its context.`;

/** Repo context worth giving the planner, cheaply and without a model call. */
function repoContext(cwd: string): string {
  const bits: string[] = [];
  const run = (cmd: string, args: string[]) => {
    try {
      return execFileSync(cmd, args, { cwd, encoding: "utf8", timeout: 5000, stdio: ["ignore", "pipe", "ignore"] }).trim();
    } catch {
      return "";
    }
  };
  const tree = run("git", ["ls-files"]);
  if (tree) {
    const files = tree.split("\n").slice(0, 200);
    bits.push(`Files (${tree.split("\n").length} tracked, first ${files.length}):\n${files.join("\n")}`);
  }
  const status = run("git", ["status", "--porcelain"]);
  if (status) bits.push(`Uncommitted changes:\n${status.slice(0, 800)}`);
  return bits.join("\n\n");
}

/**
 * Ask the planning model to decompose a task. Returns null when planning is not
 * possible or the response is unusable — the caller then runs the task directly,
 * because a failed plan must never block the work.
 */
export async function makePlan(
  task: string,
  cwd: string,
  endpoint: { baseUrl: string; apiKey?: string; model: string },
  timeoutMs = 120_000,
): Promise<Plan | null> {
  const context = repoContext(cwd);
  try {
    const res = await fetch(`${endpoint.baseUrl.replace(/\/$/, "")}/chat/completions`, {
      method: "POST",
      signal: AbortSignal.timeout(timeoutMs),
      headers: {
        "content-type": "application/json",
        "user-agent": "underclass/0.1.0-alpha.1",
        ...(endpoint.apiKey ? { authorization: `Bearer ${endpoint.apiKey}` } : {}),
      },
      body: JSON.stringify({
        model: endpoint.model,
        max_tokens: 2000,
        messages: [
          { role: "system", content: PLAN_SYSTEM },
          { role: "user", content: `${context ? `Repository:\n${context}\n\n` : ""}Task: ${task}` },
        ],
      }),
    });
    if (!res.ok) return null;
    const json = (await res.json()) as any;
    const text: string = json?.choices?.[0]?.message?.content ?? "";
    const usage = json?.usage ?? {};

    // Models wrap JSON in prose or fences no matter how firmly you ask.
    const match = text.match(/\{[\s\S]*\}/);
    if (!match) return null;
    const parsed = JSON.parse(match[0]) as { steps?: PlanStep[]; notes?: string };
    const steps = (parsed.steps ?? []).filter((s) => s && typeof s.action === "string" && s.action.trim());
    if (!steps.length) return null;

    // Reject a plan that only tells the executor to go and look.
    //
    // Measured 2026-07-25: MiniMax-M2.7 returned exactly one step — "Read
    // src/cart.js to understand current code structure" — for a two-bug task.
    // The run still succeeded, because the executor reads files anyway, but
    // planning cost 1.68x the input tokens and 2.56x the output to contribute
    // nothing. A plan that decides nothing is strictly worse than no plan: it is
    // paid for twice, once to generate and again as context on every later turn.
    //
    // Falling through to a direct run is the right failure: the task still gets
    // done, just without the tax.
    const investigation =
      /^\s*(read|open|view|inspect|examine|review|analy[sz]e|look at|locate|find|identify|understand|explore|check|study|survey)\b/i;
    const actionable = steps.filter((s) => !investigation.test(s.action));
    if (actionable.length === 0) {
      process.stderr.write(
        `\x1b[33munder: the planner returned only investigation steps ` +
          `("${steps[0]!.action.slice(0, 60)}…") — ignoring the plan and running directly.\x1b[0m\n`,
      );
      return null;
    }
    // Keep a partial plan, minus the steps that decide nothing.
    if (actionable.length < steps.length) {
      process.stderr.write(
        `\x1b[2munder: dropped ${steps.length - actionable.length} investigation step(s) from the plan\x1b[0m\n`,
      );
    }
    steps.length = 0;
    steps.push(...actionable);
    return {
      steps,
      ...(parsed.notes ? { notes: parsed.notes } : {}),
      tokensIn: usage.prompt_tokens ?? 0,
      tokensOut: usage.completion_tokens ?? 0,
      model: endpoint.model,
    };
  } catch {
    return null;
  }
}

/**
 * Render a plan as the execution prompt.
 *
 * The executor is told the plan is authoritative and the decisions are already
 * made, because a small model handed a plan will otherwise re-litigate it.
 */
export function planToPrompt(task: string, plan: Plan): string {
  const steps = plan.steps
    .map((s, i) => `${i + 1}. ${s.file ? `In \`${s.file}\`: ` : ""}${s.action}`)
    .join("\n");
  return [
    `Task: ${task}`,
    "",
    "A plan has already been made for you. Execute these steps in order, exactly as written.",
    "The decisions are made — do not re-plan, do not skip steps, do not substitute your own approach.",
    "",
    steps,
    plan.notes ? `\nContext: ${plan.notes}` : "",
    "",
    "When every step is done, stop.",
  ].join("\n");
}

File: src/planner.rs
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::process::Command;
use reqwest::Client;
use crate::agent::llm::{send_chat_completion, ChatCompletionRequest, ChatMessage};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanStep {
    pub file: Option<String>,
    pub action: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Plan {
    pub steps: Vec<PlanStep>,
    pub notes: Option<String>,
}

pub const PLAN_SYSTEM: &str = "You are planning work for a SMALL, LITERAL coding model that will execute your steps.\n\
It has file read/write/edit tools and a shell. It is not clever: it follows instructions exactly and\n\
fails when a step requires judgement or spans several files at once.\n\n\
Produce a plan as JSON only, no prose:\n\
{\"steps\":[{\"file\":\"path/to/file\",\"action\":\"one imperative instruction\"}],\"notes\":\"optional context\"}\n\n\
Rules for steps:\n\
- Each step must be independently executable and verifiable.\n\
- Name the exact file. Name the exact change. Quote the literal text to find and the literal text to replace it with wherever you can.\n\
- No step may require deciding whether to do something. You decide; the executor acts.\n\
- Prefer 1-6 steps. If the task genuinely needs more, the task is too large — say so in notes.\n\
- If a step needs a command run to verify, make that its own step with the exact command.\n\n\
NEVER emit an investigation step. Read X, examine structure, review code, analyse implementation and locate bug are all forbidden.\n\
Every step must change something or verify something.";

pub fn repo_context(cwd: &Path) -> String {
    let mut bits = Vec::new();

    if let Ok(out) = Command::new("git").args(["ls-files"]).current_dir(cwd).output() {
        if out.status.success() {
            let tree = String::from_utf8_lossy(&out.stdout);
            let lines: Vec<&str> = tree.lines().collect();
            let sample = lines.iter().take(200).copied().collect::<Vec<&str>>().join("\n");
            bits.push(format!("Files ({} tracked, first {}):\n{}", lines.len(), sample.lines().count(), sample));
        }
    }

    if let Ok(out) = Command::new("git").args(["status", "--porcelain"]).current_dir(cwd).output() {
        if out.status.success() {
            let status = String::from_utf8_lossy(&out.stdout);
            if !status.trim().is_empty() {
                let trimmed = status.chars().take(800).collect::<String>();
                bits.push(format!("Uncommitted changes:\n{trimmed}"));
            }
        }
    }

    bits.join("\n\n")
}

pub async fn generate_plan(
    client: &Client,
    base_url: &str,
    api_key: &str,
    model: &str,
    prompt: &str,
    cwd: &Path,
) -> Result<Plan, String> {
    let context = repo_context(cwd);
    let user_msg = if context.is_empty() {
        prompt.to_string()
    } else {
        format!("{prompt}\n\nRepo Context:\n{context}")
    };

    let req = ChatCompletionRequest {
        model: model.to_string(),
        messages: vec![
            ChatMessage { role: "system".to_string(), content: Some(PLAN_SYSTEM.to_string()), tool_calls: None, tool_call_id: None },
            ChatMessage { role: "user".to_string(), content: Some(user_msg), tool_calls: None, tool_call_id: None },
        ],
        tools: None,
        temperature: Some(0.1),
        max_tokens: Some(4096),
    };

    let resp = send_chat_completion(client, base_url, api_key, None, &req).await?;
    let content = resp.choices.first()
        .and_then(|c| c.message.content.clone())
        .ok_or_else(|| "Planner returned empty response".to_string())?;

    // Clean JSON codeblock wrappers if present
    let json_str = content.trim()
        .trim_start_matches("```json")
        .trim_start_matches("```")
        .trim_end_matches("```")
        .trim();

    serde_json::from_str::<Plan>(json_str).map_err(|e| format!("Failed to parse plan JSON: {e} ({json_str})"))
}

pub fn format_plan_as_prompt(plan: &Plan) -> String {
    let mut out = String::from("Execute the following plan step by step:\n");
    if let Some(ref notes) = plan.notes {
        out.push_str(&format!("Notes: {}\n\n", notes));
    }
    for (idx, step) in plan.steps.iter().enumerate() {
        if let Some(ref file) = step.file {
            out.push_str(&format!("{}. [{}] {}\n", idx + 1, file, step.action));
        } else {
            out.push_str(&format!("{}. {}\n", idx + 1, step.action));
        }
    }
    out
}
